File: scenarios/event-hubs-java-send-receive-events/golden/src/main/java/com/example/EventHubs.java
package com.example;

import com.azure.core.amqp.exception.AmqpException;
import com.azure.messaging.eventhubs.EventData;
import com.azure.messaging.eventhubs.EventDataBatch;
import com.azure.messaging.eventhubs.EventHubClientBuilder;
import com.azure.messaging.eventhubs.EventHubProducerClient;
import com.azure.messaging.eventhubs.EventProcessorClient;
import com.azure.messaging.eventhubs.EventProcessorClientBuilder;
import com.azure.messaging.eventhubs.checkpointstore.blob.BlobCheckpointStore;
import com.azure.messaging.eventhubs.models.CreateBatchOptions;
import com.azure.messaging.eventhubs.models.ErrorContext;
import com.azure.messaging.eventhubs.models.EventContext;
import com.azure.storage.blob.BlobContainerAsyncClient;
import com.azure.storage.blob.BlobContainerClientBuilder;

import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicReference;
import java.util.UUID;

public final class EventHubs {
    private EventHubs() {
    }

    public static void main(String[] args) throws InterruptedException {
        String eventHubsConnectionString =
                requireEnvironment("EVENT_HUBS_CONNECTION_STRING");
        String eventHubName = requireEnvironment("EVENT_HUB_NAME");
        String storageConnectionString =
                requireEnvironment("STORAGE_CONNECTION_STRING");
        String checkpointContainerName =
                requireEnvironment("CHECKPOINT_CONTAINER_NAME");

        try (EventHubProducerClient producer = new EventHubClientBuilder()
                .connectionString(eventHubsConnectionString, eventHubName)
                .buildProducerClient()) {
            BlobContainerAsyncClient blobContainer =
                    new BlobContainerClientBuilder()
                            .connectionString(storageConnectionString)
                            .containerName(checkpointContainerName)
                            .buildAsyncClient();
            blobContainer.createIfNotExists().block();
            BlobCheckpointStore checkpointStore =
                    new BlobCheckpointStore(blobContainer);
            CountDownLatch receivedEvents = new CountDownLatch(10);
            CountDownLatch initializedPartition = new CountDownLatch(1);
            AtomicReference<String> initializedPartitionId = new AtomicReference<>();
            String batchId = UUID.randomUUID().toString();

            EventProcessorClient processor = new EventProcessorClientBuilder()
                    .connectionString(eventHubsConnectionString, eventHubName)
                    .consumerGroup(EventHubClientBuilder.DEFAULT_CONSUMER_GROUP_NAME)
                    .checkpointStore(checkpointStore)
                    .processPartitionInitialization(context -> {
                        initializedPartitionId.compareAndSet(
                                null,
                                context.getPartitionContext().getPartitionId());
                        initializedPartition.countDown();
                    })
                    .processEvent(context ->
                            processEvent(context, receivedEvents, batchId))
                    .processError(EventHubs::processError)
                    .buildEventProcessorClient();

            try {
                processor.start();
                if (!initializedPartition.await(30, TimeUnit.SECONDS)) {
                    throw new IllegalStateException(
                            "Timed out waiting for a partition receiver to initialize.");
                }
                EventDataBatch batch = producer.createBatch(
                        new CreateBatchOptions()
                                .setPartitionId(initializedPartitionId.get()));
                for (int i = 0; i < 10; i++) {
                    EventData event = new EventData("Event " + i);
                    event.getProperties().put("eventId", i);
                    event.getProperties().put("batchId", batchId);
                    if (!batch.tryAdd(event)) {
                        throw new IllegalStateException(
                                "The ten events exceeded the batch size.");
                    }
                }
                producer.send(batch);
                if (!receivedEvents.await(30, TimeUnit.SECONDS)) {
                    throw new IllegalStateException(
                            "Timed out waiting for all ten events; received "
                                    + (10 - receivedEvents.getCount()) + ".");
                }
            } catch (AmqpException exception) {
                reportEventHubsFailure(
                        "send/receive",
                        initializedPartitionId.get(),
                        exception);
                throw exception;
            } finally {
                processor.stop();
            }
        }
    }

    private static void processEvent(
            EventContext context,
            CountDownLatch receivedEvents,
            String expectedBatchId) {
        System.out.println(context.getEventData().getBodyAsString());
        context.updateCheckpoint();
        if (expectedBatchId.equals(
                context.getEventData().getProperties().get("batchId"))) {
            receivedEvents.countDown();
        }
    }

    private static void processError(ErrorContext context) {
        Throwable throwable = context.getThrowable();
        String partitionId = context.getPartitionContext() == null
                ? "unassigned"
                : context.getPartitionContext().getPartitionId();
        if (throwable instanceof AmqpException amqpException) {
            reportEventHubsFailure("processor callback", partitionId, amqpException);
            return;
        }
        System.err.printf(
                "Event Hubs processor error: partition=%s type=%s message=%s%n",
                partitionId,
                throwable.getClass().getName(),
                throwable.getMessage());
    }

    private static void reportEventHubsFailure(
            String operation,
            String partitionId,
            AmqpException exception) {
        System.err.printf(
                "Event Hubs %s failed: partition=%s condition=%s transient=%s message=%s%n",
                operation,
                partitionId == null ? "unassigned" : partitionId,
                exception.getErrorCondition(),
                exception.isTransient(),
                exception.getMessage());
    }

    private static String requireEnvironment(String name) {
        String value = System.getenv(name);
        if (value == null || value.isBlank()) {
            throw new IllegalStateException("Set " + name + " before running.");
        }
        return value;
    }
}

File: scenarios/event-hubs-typescript-send-receive-events/golden/src/event-hubs.ts
import {
  EventHubConsumerClient,
  EventHubProducerClient,
  type Subscription,
} from "@azure/event-hubs";
import { BlobCheckpointStore } from "@azure/eventhubs-checkpointstore-blob";
import { BlobServiceClient } from "@azure/storage-blob";

declare const process: {
  env: Record<string, string | undefined>;
  once(event: "SIGINT" | "SIGTERM", listener: () => void): void;
};

function requireEnvironment(name: string): string {
  const value = process.env[name];
  if (!value) {
    throw new Error(`Set ${name} before running the application.`);
  }
  return value;
}

function waitForShutdown(): Promise<void> {
  return new Promise((resolve) => {
    process.once("SIGINT", resolve);
    process.once("SIGTERM", resolve);
  });
}

async function main(): Promise<void> {
  const connectionString = requireEnvironment("EVENT_HUB_CONNECTION_STRING");
  const eventHubName = requireEnvironment("EVENT_HUB_NAME");
  const producer = new EventHubProducerClient(connectionString, eventHubName);

  const blobServiceClient = BlobServiceClient.fromConnectionString(
    requireEnvironment("CHECKPOINT_STORAGE_CONNECTION_STRING"),
  );
  const containerClient = blobServiceClient.getContainerClient(
    requireEnvironment("CHECKPOINT_CONTAINER_NAME"),
  );
  await containerClient.createIfNotExists();
  const checkpointStore = new BlobCheckpointStore(containerClient);
  const consumer = new EventHubConsumerClient(
    EventHubConsumerClient.defaultConsumerGroupName,
    connectionString,
    eventHubName,
    checkpointStore,
  );

  let subscription: Subscription | undefined;
  try {
    const batch = await producer.createBatch();
    for (let index = 0; index < 10; index += 1) {
      const event = {
        body: `event-${index}`,
        properties: { sequence: index, source: "typescript-reference" },
      };
      if (!batch.tryAdd(event)) {
        throw new Error(`Event ${index} did not fit in the batch.`);
      }
    }
    await producer.sendBatch(batch);

    subscription = consumer.subscribe({
      processEvents: async (events, context) => {
        for (const event of events) {
          console.log(event.body);
        }
        if (events.length > 0) {
          await context.updateCheckpoint(events[events.length - 1]);
        }
      },
      processError: async (error, context) => {
        console.error("Partition failure", context.partitionId, error);
      },
    });

    await waitForShutdown();
  } finally {
    await subscription?.close();
    await consumer.close();
    await producer.close();
  }
}

await main();
